src/scram.js

```diff
@@ -0,0 +1,167 @@
+import utils from './utils';
+
+async function scramClientProof( authMessage, clientKey, hashName ) {
+    let enc = new TextEncoder();
+
+    const storedKey = await window.crypto.subtle.importKey(
+      "raw",
+      await window.crypto.subtle.digest(hashName, clientKey),
+      { "name": "HMAC", "hash": hashName },
+      false,
+      ["sign"]
+    );
+    const clientSignature = await window.crypto.subtle.sign("HMAC", storedKey, enc.encode(authMessage));
+
+    return utils.xorArrayBuffers(clientKey, clientSignature);
+}
+
+function scramParseChallenge ( challenge ) {
+    /* This function parses the information in a SASL SCRAM challenge response,
+     * into an object of the form
+     * { nonce: String,
+     *   salt:  ArrayBuffer,
+     *   iter:  Int
+     * }
+     * Returns undefined on failure.
+     */
+    let nonce, salt, iter;
+    const attribMatch = /([a-z]+)=([^,]+)(,|$)/;
+    while (challenge.match(attribMatch)) {
+        const matches = challenge.match(attribMatch);
+        challenge = challenge.replace(matches[0], "");
+        switch (matches[1]) {
+        case "r":
+            nonce = matches[2];
+            break;
+        case "s":
+            salt = utils.base64ToArrayBuf(matches[2]);
+            break;
+        case "i":
+            iter = parseInt(matches[2]);
+            break;
+        default: return undefined;
+        }
+    }
+
+    // Consider iteration counts less than 4096 insecure, as reccommended by
+    // RFC 5802
+    if (isNaN(iter) || iter < 4096) {
+        return undefined;
```

Comment:
Could you please log the fact that this challenge is ignored as insecure?

For example:
```
Strophe.warn("Ignored insecure SASL SCRAM challenge with less than 4096 iteration counts");
```

Reply:
You might have an issue with getting `Strophe.warn` here since it looks like it would result in circular imports.

In that case, you can move the log functions into a separate file, e.g. `log.js` and import them here (and in core.js wher e you add them to the `Strophe` object).

Reply:
After looking at this, I think this would probably break code out in the wild which relies on being able to set Strophe.log/warn/error to custom logging/error routines. A quick fix would just be to make the `scramResponse` function take a logging object as an argument, populated with these functions like `{ log: f(x), warn: g(x), ... }`. I don't know if this is how you want to architect your program though, it seems a bit messy to me.
Realistically though, the probability that this will ever actually happen in the wild is pretty close to 0. I don't think any legitimate xmpp server still running would suggest an iteration count under 4096 iterations, or supply an incorrect salt. We might be fine with just using good old `console.warn` for this, then just mark it with big red tape saying that its bad and we'll Fix it Eventually™. Let me know what you think.

Reply:
I have an idea that I'd like to explore, just give me a bit more time please.

Reply:
If this is still holding up the pull, we could always just return an object containing the error string and error severity instead of undefined, and just check for the error string in `_sasl_challenge_cb`, then just use the logging functions there. It's out of style for the rest of the project, but it's probably not that big of a deal.